File: src/tools/listDirectory.ts
import { readdirSync, statSync } from 'node:fs';
import { join } from 'node:path';
import { resolveInProject } from '../safety/policy.ts';
import { relFrom } from '../project/walker.ts';
import { isIgnored, type IgnoreRule, loadGitignoreRules } from '../project/gitignore.ts';
import { formatBytes } from '../util/text.ts';
import type { Tool, ToolContext } from './registry.ts';

export function createListDirectoryTool(ctx: ToolContext): Tool {
  return {
    name: 'list_directory',
    description: 'List files/dirs in a project directory.',
    parameters: {
      type: 'object',
      properties: {
        path: { type: 'string', description: 'Dir relative to root (default root).' },
        depth: { type: 'number', description: 'Recursion depth (1-5, default 1).' },
        recursive: { type: 'boolean', description: 'Shorthand for depth 3.' },
      },
    },
    async execute(args) {
      const rawPath = typeof args.path === 'string' ? args.path : '';
      let depth = typeof args.depth === 'number' && Number.isFinite(args.depth) ? Math.floor(args.depth) : 1;
      if (args.recursive === true && depth <= 1) depth = 3;
      depth = Math.max(1, Math.min(depth, 5));

      const resolved = rawPath ? resolveInProject(ctx.projectRoot, rawPath) : { abs: ctx.projectRoot, real: ctx.projectRoot, ok: true as const };
      if (!resolved.ok) return `Error: ${resolved.reason}`;
      if (!statSyncSafe(resolved.abs)?.isDirectory()) {
        return `Error: "${rawPath || '.'}" is not a directory inside the project.`;
      }

      const lines: string[] = [];
      const rootRules = loadGitignoreRules(ctx.projectRoot, '');
      walk(resolved.abs, 0, rootRules);

      function walk(dir: string, level: number, rules: IgnoreRule[]): void {
        if (level > depth) return;
        let entries: string[];
        try {
          entries = readdirSync(dir).sort();
        } catch {
          return;
        }
        const relDir = dir === ctx.projectRoot ? '' : relFrom(ctx.projectRoot, dir);
        const nested = dir === ctx.projectRoot ? rules : loadGitignoreRules(dir, relDir);
        const frameRules = rules.concat(nested);

        for (const name of entries) {
          const abs = join(dir, name);
          const rel = relDir ? `${relDir}/${name}` : name;
          let st;
          try {
            st = statSync(abs);
          } catch {
            continue;
          }
          const isDir = st.isDirectory();
          // Respect .gitignore (skips .git, node_modules, .myagent, etc.)
          if (isIgnored(rel, isDir, frameRules)) continue;

          const indent = '  '.repeat(level);
          if (isDir) {
            lines.push(`${indent}${name}/`);
            walk(abs, level + 1, frameRules);
          } else {
            lines.push(`${indent}${name}  ${formatBytes(st.size)}`);
          }
          if (lines.length > 500) {
            lines.push('… [too many entries, listing truncated]');
            return;
          }
        }
      }

      if (lines.length === 0) return '(empty directory)';
      const label = rawPath ? relFrom(ctx.projectRoot, resolved.abs) : '(project root)';
      return `Listing ${label}:\n${lines.join('\n')}`;
    },
  };
}

function statSyncSafe(p: string): ReturnType<typeof statSync> | null {
  try {
    return statSync(p);
  } catch {
    return null;
  }
}

File: grace/tools/list_directory.py
"""list_directory tool (port of src/tools/listDirectory.ts)."""

import os

from grace.project.gitignore import is_ignored, load_gitignore_rules
from grace.project.walker import rel_from
from grace.safety import resolve_in_project
from grace.tools.tool import Tool
from grace.util_text import format_bytes


def create_list_directory_tool(ctx) -> Tool:
    def execute(args: dict, tool_ctx) -> str:
        raw_path = args.get("path") if isinstance(args.get("path"), str) else ""
        depth = 1
        if isinstance(args.get("depth"), (int, float)) and depth is not None:
            depth = int(args.get("depth"))
        if args.get("recursive") is True and depth <= 1:
            depth = 3
        depth = max(1, min(depth, 5))

        if raw_path:
            resolved = resolve_in_project(ctx.projectRoot, raw_path)
            if not resolved["ok"]:
                return f"Error: {resolved['reason']}"
        else:
            resolved = {"abs": ctx.projectRoot, "real": ctx.projectRoot, "ok": True}

        try:
            if not os.path.isdir(resolved["abs"]):
                return f'Error: "{raw_path or "."}" is not a directory inside the project.'
        except OSError:
            return f'Error: "{raw_path or "."}" is not a directory inside the project.'

        lines: list[str] = []
        root_rules = load_gitignore_rules(ctx.projectRoot, "")

        def walk(dir_path: str, level: int, rules: list) -> None:
            if level > depth:
                return
            try:
                entries = sorted(os.listdir(dir_path))
            except OSError:
                return
            rel_dir = "" if dir_path == ctx.projectRoot else rel_from(ctx.projectRoot, dir_path)
            nested = rules if dir_path == ctx.projectRoot else load_gitignore_rules(dir_path, rel_dir)
            frame_rules = rules + nested

            for name in entries:
                abs_path = os.path.join(dir_path, name)
                rel = f"{rel_dir}/{name}" if rel_dir else name
                try:
                    st = os.stat(abs_path)
                except OSError:
                    continue
                is_dir = os.path.isdir(abs_path)
                # Respect .gitignore (skips .git, node_modules, .zeesh, .myagent, etc.)
                if is_ignored(rel, is_dir, frame_rules):
                    continue

                indent = "  " * level
                if is_dir:
                    lines.append(f"{indent}{name}/")
                    walk(abs_path, level + 1, frame_rules)
                else:
                    lines.append(f"{indent}{name}  {format_bytes(st.st_size)}")
                if len(lines) > 500:
                    lines.append("… [too many entries, listing truncated]")
                    return

        walk(resolved["abs"], 0, root_rules)

        if not lines:
            return "(empty directory)"
        label = rel_from(ctx.projectRoot, resolved["abs"]) if raw_path else "(project root)"
        return f"Listing {label}:\n" + "\n".join(lines)

    return Tool(
        name="list_directory",
        description="List files/dirs in a project directory.",
        parameters={
            "type": "object",
            "properties": {
                "path": {"type": "string", "description": "Dir relative to root (default root)."},
                "depth": {"type": "number", "description": "Recursion depth (1-5, default 1)."},
                "recursive": {"type": "boolean", "description": "Shorthand for depth 3."},
            },
        },
        execute=execute,
    )
